### ISSUE TITLE
IllegalMonitorStateException in while loop with wait()

### ISSUE THREAD

**Issue by shannah** [konsoletyper/teavm#92]
I'm running into an IllegalMonitorStateException in a demo app when it shouldn't be (at least I'm 99.999% sure it shouldn't be).

I've uploaded the app here
https://github.com/shannah/cn1-teavmport-demos/tree/master/DrSbaitso

The exception occurs on the first screen when you click the "Take Photo" button.

![screen shot 2015-03-27 at 9 27 57 am](https://cloud.githubusercontent.com/assets/2677562/6872102/22897f88-d464-11e4-9256-9c98d6eb6d0d.png)

I believe that the error is happening in relation to this while loop:

~~~
while(!completed) {
                try {
                    wait();
                } catch (InterruptedException ex) {
                }
            }
~~~

https://github.com/shannah/cn1/blob/master/CodenameOne/src/com/codename1/capture/Capture.java#L166

**Comment by shannah** [comment]
I should add, in case you're trying to reproduce this.  The "Take Photo" button won't work right now as I haven't implemented the native portions.  But it shouldn't run into the monitor state exception.  It should either do nothing (or possibly return a different exception).  I say this just so you don't think it's another bug when it doesn't work :)

**Comment by konsoletyper** [comment]
Actually, it's not a bug, it's just lack of a feature. TeaVM does not support `synchronized` methods. Will implement support as soon as possible.

**Comment by shannah** [comment]
Ahh.. That makes sense.  I didn't realize that synchronized methods didn't use monitorEnter/Exit.  Just read the JVM spec on the topic.

**Comment by shannah** [comment]
Not yet.  I temporarily changed problematic synchronized methods in the app to use synchronized blocks as a workaround. I see that you merged it to master though so I'll be trying that out tomorrow or Tuesday.  Thanks for all the work on this.

**Comment by shannah** [comment]
Just getting around to this now.  Support for synchronized methods is flushing out a few `Method *.<clinit>()V is claimed to be synchronous, but it is has invocations of asynchronous methods` errors, and this is cascading down... "

I'm experimenting with flags to "disable synchronization" for sections of code to make this easier.
E.g.
At the beginning of clinit()
Thread.disableSynchronization()

and at the end of clinit()
Thread.enableSynchronization()

And just let the pieces fall where they may.    I know this isn't a perfect solution.

**Comment by konsoletyper** [comment]
Don't waste your time. TeaVM is too pessimistic giving these error messages. Try `<stopOnErrors>false</stopOnErrors>` in Maven plugin configuration. I'll try to implement asynchronous `<clinint>` methods, since it's a bit easier with `switch`-based CPS than with initial callback-based.

**Comment by shannah** [comment]
I've already "wasted" a little time, and have it *almost* building.  Obviously a solution that supports async inside clinit would be better though.

One other issue I've run into with relation to the new synchronized method support is that any use of an interface for which here exists at least one sub class with a synchronized version of a method seems to flag the code as async.  This is problematic in places where you can't have async code (e.g. inside `@JSFunctor` methods. 

E.g. I have implemented `requestAnimationFrame` which cannot have async code, and would not function properly if I did introduce async code with the following:

~~~
animationFrameCallback = new AnimationFrameCallback(){

            @Override
            public void onAnimationFrame(int time) {
                 ...
                List<ExecutableOp> ops = new ArrayList<ExecutableOp>();
                
                ops.addAll(pendingDisplay.ops);
                ... 
            }
            
        };
~~~

This causes a problem because the call to ops.addAll() is on the List interface which is implemented by some synchronized classes.

I thought I might get around this problem by explicitly declaring `ops` to be of type `ArrayList`, e.g.:

~~~
ArrayList<ExecutableOp> ops = new ArrayList<ExecutableOp>();
~~~

But the `ops.addAll()` call is still problematic because its signature takes a `Collection` as parameter, so its internal code doesn't "know" whether the collection is synchronized or not - and thus addAll() is flagged as Async.

I'll give `stopOnErrors=false` a shot and see if this clears things up.

**Comment by konsoletyper** [comment]
From @Functors you can launch a new thread and execute any async code there.

**Comment by shannah** [comment]
Yes.  In the case of `requestAnimationFrame` the point of that method is to ensure that the "frame" is ready to be displayed for the coming screen paint, so if I wrap it in a new Thread, it may defer the execution which may cause problems.  (Haven't tried it, but conceptually it seems problematic).

**Comment by konsoletyper** [comment]
Sorry, TeaVM can't build more precise call graph. And with this call graph it produces lots of false positives when using with code that wasn't initially designed for TeaVM. Yes, it may be reasonable to invent annotation (not these strange methods) that suppresses errors on your own risk. Can `requestAnimationFrame` post task to a CN1's own event loop?

**Comment by shannah** [comment]
With `stopOnErrors=false` it seems to be OK so far.  It reports the errors, but the app still runs... this seems like an acceptable compromise.
